File: backend/internal/handler/calendar.go
package handler

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	"github.com/thienduchuutran/church-website/backend/internal/middleware"
	"github.com/thienduchuutran/church-website/backend/internal/model"
	"github.com/thienduchuutran/church-website/backend/internal/service"
)

type CalendarHandler struct {
	svc *service.CalendarService
}

func NewCalendarHandler(svc *service.CalendarService) *CalendarHandler {
	return &CalendarHandler{svc: svc}
}

// GetMonth handles GET /api/v1/calendar?year=2026&month=4&locale=vi
func (h *CalendarHandler) GetMonth(w http.ResponseWriter, r *http.Request) {
	year, month, ok := parseYearMonth(w, r)
	if !ok {
		return
	}
	// An absent ?locale= means English, not "raw". Before migration 000013 the two
	// were the same thing - the source column WAS English - so the frontend omits
	// the param for English viewers as a small optimization. Now they differ: the
	// raw path skips translation entirely, which would serve a Vietnamese-authored
	// event untranslated to an English reader. Defaulting here rather than
	// requiring the client to send it keeps old clients and direct API calls
	// correct, and leaves the raw path reachable only from internal service calls
	// that genuinely want stored text.
	locale := strings.TrimSpace(r.URL.Query().Get("locale"))
	if locale == "" {
		locale = "en"
	}
	resp, err := h.svc.GetMonth(r.Context(), year, month, locale)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to load calendar")
		return
	}
	// Admin-only fields, all stripped in one place so there is a single boundary
	// to audit:
	//   - private_address: non-admins only see an address explicitly marked
	//     public on the site.
	//   - place / place_id: the venue that address resolved to. Stripped under
	//     the SAME condition, because a place name identifies a household as
	//     precisely as its street number - leaving "MST House" behind while
	//     hiding "203 Essex Street" would defeat the whole point of the flag.
	//   - title_source / notes_source / content_source: the untranslated English
	//     text, needed only so the admin edit form saves the source instead of
	//     the machine translation it is displaying. A public visitor has no use
	//     for it, and shipping it would double the text in every payload.
	if middleware.AdminEmailFromContext(r.Context()) == "" {
		for i := range resp.Events {
			if !resp.Events[i].AddressPublic {
				resp.Events[i].PrivateAddress = nil
				resp.Events[i].Place = nil
				resp.Events[i].PlaceID = nil
			}
			resp.Events[i].TitleSource = nil
			resp.Events[i].NotesSource = nil
		}
		if resp.MonthNote != nil {
			resp.MonthNote.ContentSource = nil
		}
	}
	writeJSON(w, http.StatusOK, resp)
}

// CreateEvent handles POST /api/v1/calendar/events (admin only).
func (h *CalendarHandler) CreateEvent(w http.ResponseWriter, r *http.Request) {
	var req model.CreateCalendarEventRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	userID := middleware.UserIDFromContext(r.Context())
	e, err := h.svc.CreateEvent(r.Context(), req, userID)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	writeJSON(w, http.StatusCreated, e)
}

// UpdateEvent handles PATCH /api/v1/calendar/events/{id} (admin only).
func (h *CalendarHandler) UpdateEvent(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	var req model.UpdateCalendarEventRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	e, err := h.svc.UpdateEvent(r.Context(), id, req)
	if err != nil {
		if errors.Is(err, model.ErrNotFound) {
			writeError(w, http.StatusNotFound, "event not found")
			return
		}
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, e)
}

// DeleteEvent handles DELETE /api/v1/calendar/events/{id} (admin only).
func (h *CalendarHandler) DeleteEvent(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	if err := h.svc.DeleteEvent(r.Context(), id); err != nil {
		if errors.Is(err, model.ErrNotFound) {
			writeError(w, http.StatusNotFound, "event not found")
			return
		}
		writeError(w, http.StatusInternalServerError, "failed to delete event")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// ListPlaces handles GET /api/v1/calendar/places (admin only).
//
// Unlike /calendar/event-types and /calendar/palette - deliberately public
// lists of labels and hex codes - this one returns addresses, including every
// address never marked address_public. It must stay inside the RequireAdmin
// group; see the route comments in cmd/server/main.go.
func (h *CalendarHandler) ListPlaces(w http.ResponseWriter, r *http.Request) {
	places, err := h.svc.ListPlaces(r.Context())
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to load places")
		return
	}
	writeJSON(w, http.StatusOK, places)
}

// RenamePlace handles PATCH /api/v1/calendar/places/{id} (admin only).
//
// Renaming pins the label against the naming model, so this is how a wrong
// model answer gets corrected - and one call fixes every event at the address.
func (h *CalendarHandler) RenamePlace(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	var req model.UpdateCalendarPlaceRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	p, err := h.svc.RenamePlace(r.Context(), id, req)
	if err != nil {
		if errors.Is(err, model.ErrNotFound) {
			writeError(w, http.StatusNotFound, "place not found")
			return
		}
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, p)
}

// ListEventTypes handles GET /api/v1/calendar/event-types.
// Public, like GET /calendar: the category vocabulary is not sensitive, and the
// day modal needs the labels to render an event's category to visitors.
func (h *CalendarHandler) ListEventTypes(w http.ResponseWriter, r *http.Request) {
	types, err := h.svc.ListEventTypes(r.Context())
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to load event types")
		return
	}
	writeJSON(w, http.StatusOK, types)
}

// CreateEventType handles POST /api/v1/calendar/event-types (admin only).
// Behaves as get-or-create, so an admin who re-types an existing label gets
// that type back with 200 rather than a duplicate error.
func (h *CalendarHandler) CreateEventType(w http.ResponseWriter, r *http.Request) {
	var req model.CreateEventTypeRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	userID := middleware.UserIDFromContext(r.Context())
	t, err := h.svc.CreateEventType(r.Context(), req, userID)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	writeJSON(w, http.StatusCreated, t)
}

// ListPaletteColors handles GET /api/v1/calendar/palette.
// Public for the same reason as the event types - it is a list of hex strings,
// and keeping it ungated means the picker has one less auth path to get wrong.
func (h *CalendarHandler) ListPaletteColors(w http.ResponseWriter, r *http.Request) {
	colors, err := h.svc.ListPaletteColors(r.Context())
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to load palette")
		return
	}
	writeJSON(w, http.StatusOK, colors)
}

// CreatePaletteColor handles POST /api/v1/calendar/palette (admin only).
func (h *CalendarHandler) CreatePaletteColor(w http.ResponseWriter, r *http.Request) {
	var req model.CreatePaletteColorRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	userID := middleware.UserIDFromContext(r.Context())
	c, err := h.svc.CreatePaletteColor(r.Context(), req, userID)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	writeJSON(w, http.StatusCreated, c)
}

// DeletePaletteColor handles DELETE /api/v1/calendar/palette/{id} (admin only).
func (h *CalendarHandler) DeletePaletteColor(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	if err := h.svc.DeletePaletteColor(r.Context(), id); err != nil {
		if errors.Is(err, model.ErrNotFound) {
			writeError(w, http.StatusNotFound, "color not found")
			return
		}
		writeError(w, http.StatusInternalServerError, "failed to delete color")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// UpsertMonthNote handles PUT /api/v1/calendar/months/{year}/{month}/note (admin only).
func (h *CalendarHandler) UpsertMonthNote(w http.ResponseWriter, r *http.Request) {
	yearStr := chi.URLParam(r, "year")
	monthStr := chi.URLParam(r, "month")
	year, err1 := strconv.Atoi(yearStr)
	month, err2 := strconv.Atoi(monthStr)
	if err1 != nil || err2 != nil || month < 1 || month > 12 {
		writeError(w, http.StatusBadRequest, "invalid year or month")
		return
	}
	var req model.UpsertMonthNoteRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	userID := middleware.UserIDFromContext(r.Context())
	note, err := h.svc.UpsertMonthNote(r.Context(), year, month, req, userID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "failed to save month note")
		return
	}
	writeJSON(w, http.StatusOK, note)
}

// UpsertMonthSettings handles PUT /api/v1/calendar/months/{year}/{month}/settings (admin only).
// Stores the accent color the admin picked in the inline color picker; the GET
// /calendar response surfaces it back so every visitor sees the same tint.
func (h *CalendarHandler) UpsertMonthSettings(w http.ResponseWriter, r *http.Request) {
	yearStr := chi.URLParam(r, "year")
	monthStr := chi.URLParam(r, "month")
	year, err1 := strconv.Atoi(yearStr)
	month, err2 := strconv.Atoi(monthStr)
	if err1 != nil || err2 != nil || month < 1 || month > 12 {
		writeError(w, http.StatusBadRequest, "invalid year or month")
		return
	}
	var req model.UpsertMonthSettingsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	userID := middleware.UserIDFromContext(r.Context())
	settings, err := h.svc.UpsertMonthSettings(r.Context(), year, month, req, userID)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, settings)
}

func parseYearMonth(w http.ResponseWriter, r *http.Request) (int, int, bool) {
	q := r.URL.Query()
	yearStr := q.Get("year")
	monthStr := q.Get("month")
	if yearStr == "" || monthStr == "" {
		writeError(w, http.StatusBadRequest, "year and month query params are required")
		return 0, 0, false
	}
	year, err1 := strconv.Atoi(yearStr)
	month, err2 := strconv.Atoi(monthStr)
	if err1 != nil || err2 != nil || month < 1 || month > 12 {
		writeError(w, http.StatusBadRequest, "invalid year or month")
		return 0, 0, false
	}
	return year, month, true
}

File: frontend/lib/calendar.ts
import { apiDelete, apiGet, apiPatch, apiPost, apiPut } from './api'
import type {
  CalendarEvent,
  CalendarEventType,
  CalendarEventTypeDef,
  CalendarMonthNote,
  CalendarMonthResponse,
  CalendarMonthSettings,
  CalendarPlace,
  CalendarSeries,
  PaletteColor,
  RecurrenceRule,
  WriteScope,
} from '@/components/features/calendar/types'

const BASE = '/api/v1/calendar'

// getMonth accepts an optional locale so viewers see translated event titles,
// notes, and the month note. EVERY viewer passes the locale they picked, admins
// included - the calendar is a display surface, and an admin on /vi should see
// what the congregation sees.
//
// The accessToken is a separate concern: when present it opts the request into
// the OptionalAdmin middleware path, which reveals admin-only fields -
// private_address, plus title_source/notes_source/content_source carrying the
// untranslated English that EventModal pre-fills from. That pairing is what lets
// an admin view Vietnamese and still save English; see CalendarShell.
//
// Note this deliberately differs from lib/posts.ts, where admin call sites omit
// the locale. Those are edit surfaces (the dashboard list feeds the edit modal
// directly), so they want the source in the display field itself.
export async function getMonth(
  year: number,
  month: number,
  accessToken?: string | null,
  locale?: string,
): Promise<CalendarMonthResponse> {
  const params = new URLSearchParams({ year: String(year), month: String(month) })
  if (locale && locale !== 'en') params.set('locale', locale)
  return apiGet(`${BASE}?${params.toString()}`, accessToken) as Promise<CalendarMonthResponse>
}

// No language field on any write. The backend detects the source language from
// the submitted text (majority of words wins) and sets source_locale itself, so
// there is nothing for a client to declare or get wrong.
export async function createEvent(
  payload: {
    date: string
    end_date?: string | null
    title: string
    event_type: CalendarEventType
    icon: string
    color: string
    private_address?: string | null
    address_public?: boolean
    notes: string | null
    // Turns this create into a series. Omitted for a one-off. The occurrences
    // are generated server-side - the client never computes dates and never
    // sends more than this single event.
    recurrence?: RecurrenceRule | null
    // The admin's "Ends on" date. Omitted means open-ended, which is the only
    // case that ever needs extending later.
    recurrence_until?: string | null
  },
  token: string,
): Promise<CalendarEvent> {
  return apiPost(`${BASE}/events`, payload, token) as Promise<CalendarEvent>
}

export async function updateEvent(
  id: string,
  payload: {
    title?: string
    event_type?: CalendarEventType
    icon?: string
    color?: string
    private_address?: string | null
    address_public?: boolean
    notes?: string | null
    // Always sent on edit (a date string or null) because the backend writes
    // end_date directly - omitting it would clear an existing span.
    end_date?: string | null
    // Changes how the series repeats. Send null to leave the rule untouched;
    // an empty string means "stop repeating", which keeps the dates already on
    // the calendar and only stops generating more. Accepted only with
    // scope=series, because a new rule rebuilds the whole series' future.
    recurrence?: RecurrenceRule | null
    recurrence_until?: string | null
    // Required when recurrence is set to '' (turning repeating off), rejected
    // otherwise. 'keep' leaves every generated date alone; 'future' also
    // removes the ones that have not happened yet. The backend refuses a
    // missing value rather than choosing - "stop repeating" has two reasonable
    // meanings and only the admin knows which one they meant.
    recurrence_cleanup?: 'keep' | 'future' | null
  },
  scope: WriteScope,
  token: string,
): Promise<CalendarEvent> {
  return apiPatch(`${BASE}/events/${id}?scope=${scope}`, payload, token) as Promise<CalendarEvent>
}

// scope is required, not optional, and is a query parameter on both the edit
// and the delete. The backend rejects a request without one instead of picking
// a default, so there is no call site that can quietly mean "all events" when
// it meant "this one".
export async function deleteEvent(id: string, scope: WriteScope, token: string): Promise<void> {
  await apiDelete(`${BASE}/events/${id}?scope=${scope}`, token)
}

// --- Recurring series (the admin's view of what repeats) ---

// Admin-only, like getPlaces: it describes the calendar's internals rather than
// its contents, and lists every series whether or not its events are public.
export async function getSeries(token: string): Promise<CalendarSeries[]> {
  return apiGet(`${BASE}/series`, token) as Promise<CalendarSeries[]>
}

// Tops a series up by another horizon's worth of dates. Purely additive - it
// only ever appends future occurrences, never moves or removes one - which is
// why it takes no scope and needs no confirmation, unlike every other write on
// the calendar.
export async function extendSeries(id: string, token: string): Promise<{ added: number }> {
  return apiPost(`${BASE}/series/${id}/extend`, {}, token) as Promise<{ added: number }>
}

// --- Places (the venue registry behind the Locations strip) ---

// Admin-only, unlike getEventTypes/getPaletteColors below. Those are public
// lists of labels and hex codes; this one returns street addresses including
// every address never marked "show on website", so the token is required rather
// than optional.
export async function getPlaces(token: string): Promise<CalendarPlace[]> {
  return apiGet(`${BASE}/places`, token) as Promise<CalendarPlace[]>
}

// Renames a venue. One call relabels every event at that address, because the
// name lives on the place rather than on each event - and it pins the label so
// the naming model can never overwrite it again.
export async function renamePlace(id: string, name: string, token: string): Promise<CalendarPlace> {
  return apiPatch(`${BASE}/places/${id}`, { name }, token) as Promise<CalendarPlace>
}

// --- Event types (the admin-managed category vocabulary) ---

// Public read - the day modal needs the labels to name an event's category to
// visitors, so this deliberately takes no token.
export async function getEventTypes(): Promise<CalendarEventTypeDef[]> {
  return apiGet('/api/v1/calendar/event-types') as Promise<CalendarEventTypeDef[]>
}

// Creates a reusable event type from an admin-typed label. The slug is derived
// server-side from the label, which makes this get-or-create: two admins who
// both type "Baptism" converge on one type instead of near-duplicates.
export async function createEventType(
  payload: { label: string; default_icon: string; default_color: string },
  token: string,
): Promise<CalendarEventTypeDef> {
  return apiPost('/api/v1/calendar/event-types', payload, token) as Promise<CalendarEventTypeDef>
}

// --- Palette colors (the shared custom swatch grid) ---

export async function getPaletteColors(): Promise<PaletteColor[]> {
  return apiGet('/api/v1/calendar/palette') as Promise<PaletteColor[]>
}

// Saves a swatch for every admin to reuse. Idempotent server-side, so adding a
// color that is already saved returns the existing swatch rather than erroring.
export async function createPaletteColor(hex: string, token: string): Promise<PaletteColor> {
  return apiPost('/api/v1/calendar/palette', { hex }, token) as Promise<PaletteColor>
}

// Removes a swatch from the picker only. Events already using that hex keep it -
// the color is copied onto the event, never referenced - so this never repaints
// the calendar.
export async function deletePaletteColor(id: string, token: string): Promise<void> {
  await apiDelete(`/api/v1/calendar/palette/${id}`, token)
}

// The four fields an admin edits together in the month modal. Passed as an
// object rather than four positional strings, because four same-typed
// parameters in a row is a call-site bug waiting to happen - swapping the verse
// and its reference would type-check perfectly.
export interface MonthNoteInput {
  content: string
  theme: string
  verse_text: string
  verse_reference: string
  /**
   * The same verse in the other language, filed as a human-authored, already
   * approved translation rather than queued for the model.
   *
   * The month modal does not currently collect it - the verse shows in the
   * language it was written in on both locales. The backend still accepts it,
   * so offering a second verse box again is a UI-only change. Omitting the key
   * clears any wording previously stored for the other locale.
   */
  verse_text_alt?: string
}

export async function upsertMonthNote(
  year: number,
  month: number,
  fields: MonthNoteInput,
  token: string,
): Promise<CalendarMonthNote> {
  return apiPut(`${BASE}/months/${year}/${month}/note`, fields, token) as Promise<CalendarMonthNote>
}

export async function upsertMonthSettings(
  year: number,
  month: number,
  accentColor: string,
  token: string,
): Promise<CalendarMonthSettings> {
  return apiPut(
    `${BASE}/months/${year}/${month}/settings`,
    { accent_color: accentColor },
    token,
  ) as Promise<CalendarMonthSettings>
}
